doc/extract.1
.TH EXTRACT 1 "Aug 7, 2012" "libextractor 1.0.0"
.\" $Id
.SH NAME
extract
\- determine meta-information about a file
.SH SYNOPSIS
.B extract
[
.B \-bgihLmnvV
]
[
.B \-l
.I library
]
[
.B \-p
.I type
]
[
.B \-x
.I type
]
.I file
\&...
.br
.SH DESCRIPTION
This manual page documents version 1.0.0 of the
.B extract
command.
.PP
.B extract
tests each file specified in the argument list in an attempt to infer meta\-information from it.  Each file is subjected to the meta\-data extraction libraries from
.I libextractor.
.PP
libextractor classifies meta\-information (also referred to as keywords) into types. A list of all types can be obtained with the
.B \-L
option.

.SH OPTIONS
.TP 8
.B \-b
Display the output in BiBTeX format.
.TP 8
.B \-g
Use grep\-friendly output (all keywords on a single line for each file).  Use the verbose option to print the filename first, followed by the keywords.  Use the verbose option twice to also display the keyword types.  This option will not print keyword types or non\-textual metadata.
.TP 8
.B \-h
Print a brief summary of the options.
.TP 8
.B \-i
Run plugins in\-process (for debugging).  By default, each plugin is run in its own process.
.TP 8
.BI \-l " libraries"
Use the specified libraries to extract keywords. The general format of libraries is .I [[\-]LIBRARYNAME[:[\-]LIBRARYNAME]*] where LIBRARYNAME is a libextractor compatible library and typically of the form .Ijpeg\. The minus before the libraryname indicates that this library should be removed from the existing list.  To run only a few selected plugins, use \-l in combination with \-n.
.TP 8
.B \-L
Print a list of all known keyword types.
.TP 8
.B \-m
Load the file into memory and perform extraction from memory (for debugging).
.TP 8
.B \-n
Do not use the default set of extractors (typically all standard extractors, currently mp3, ogg, jpg, gif, png, tiff, real, html, pdf and mime\-types), use only the extractors specified with the .B \-l option.
.TP
.B \-p " type"
Print only the keywords matching the specified type. By default, all keywords that are found and not removed as duplicates are printed.
.TP 8
.B \-v
Print the version number and exit.
.TP 8
.B \-V
Be verbose.  This option can be specified multiple times to increase verbosity further.
.TP 8
.I \-x " type"
Exclude keywords of the specified type from the output. By default, all keywords that are found and not removed as duplicates are printed.
.SH "SEE ALSO"
.BR libextractor (3)
\- description of the libextractor library
.br
.SH EXAMPLES
.nf
$ extract test/test.jpg
comment \- (C) 2001 by Christian Grothoff, using gimp 1.2 1
mimetype \- image/jpeg

$ extract \-V \-x comment test/test.jpg
Keywords for file test/test.jpg:
mimetype \- image/jpeg

$ extract \-p comment test/test.jpg
comment \- (C) 2001 by Christian Grothoff, using gimp 1.2 1

$ extract \-nV \-l png.so \-p comment test/test.jpg test/test.png
Keywords for file test/test.jpg:
Keywords for file test/test.png:
comment \- Testing keyword extraction

.SH "LEGAL NOTICE"
libextractor and the extract tool are released under the GPL.  libextractor is a GNU package.

.SH BUGS
A couple of file\-formats (on the order of 10^3) are not recognized...

.SH AUTHORS
.B extract
was originally written by Christian Grothoff <christian@grothoff.org> and Vidyut Samanta <vids@cs.ucla.edu>. Use <libextractor@gnu.org> to contact the current maintainer(s).

.SH AVAILABILITY
You can obtain the original author's latest version from http://www.gnu.org/software/libextractor/


NEWS
Fri Aug 17 21:28:33 CEST 2012
    	Major changes to plugin mechanism now allow out-of-process
	plugins full random-access to the entire file.  Most plugins
	have now been rewritten to the new plugin API (0.7).  The external
	(libextractor) API will remain unchanged and compatible with LE 0.6.

	As part of the rewrite, many plugins were changed to use 
	standard 3rd party libraries (libjpeg, libtiff, libgif, 
	libtidy, libmagic) for parsing.  A new plugin based on
	gstreamer replaces many existing multimedia plugins.

	Automated testcases for (almost all) of the plugins
	were also written and the documentation was updated.

	The license was changed to GNU GPL v3+.  
	PDF support removed (GPLv2-only).
	REAL support removed (not yet ported).

Mon Nov 28 12:17:42 CET 2011
    	Added MKV (Matroska) plugin.
	Releasing libextractor 0.6.3.

Sun Jun 13 13:15:43 CEST 2010
	Releasing libextractor 0.6.2.

Sun Mar 14 00:09:45 CET 2010
	Releasing libextractor 0.6.1.

Wed Jan 13 17:11:07 CET 2010
        Major API change (no longer returning linked lists, using callbacks instead).
	Releasing libextractor 0.6.0.

Sat Jul  4 23:05:22 CEST 2009
	Releasing libextractor 0.5.23.

Fri Feb 20 00:28:00 MST 2009
	Releasing libextractor 0.5.22.

Sun Nov  2 20:19:02 MST 2008
	Releasing libextractor 0.5.21.

Tue Aug 12 04:40:49 EEST 2008
	Added an S3M (Scream Tracker 3 Module) plugin.
	Added an XM (eXtended Module) plugin.
	Added an IT (Impulse Tracker) plugin.

Mon Jun 23 19:05:07 EET 2008
	Fixed concurrency issues in plugin (un-)loading by adding locking
	around libltdl functions.

Fri Jun 20 23:34:02 EET 2008
	Added an FFmpeg-based thumbnail extractor plugin, initially
	supporting only bmp and png files.

Mon Mar 21 00:00:52 MDT 2008
	Releasing libextractor 0.5.20a.

Thu Mar 20 23:38:47 MDT 2008
	Releasing libextractor 0.5.20.

Wed Feb 13 10:16:55 EET 2008
	Added a plugin for AppleSingle/AppleDouble files.

Sat Jan 12 14:14:32 MST 2008
	Releasing libextractor 0.5.19a.

Mon Jan  7 08:46:40 MST 2008
	Releasing libextractor 0.5.19.

Wed Dec 26 19:38:22 MST 2007
	Added a FLAC (.flac) plugin.

Wed Dec 26 14:50:11 EET 2007
	Added a Flash Video (.flv) plugin.

Sat Jan  6 14:27:18 EET 2007
	Added an NSFE (Extended Nintendo Sound Format) plugin.

Mon Jan  1 19:10:40 MST 2007
	Added pkgconfig support.  Releasing libextractor 0.5.17.

Mon Nov 20 22:08:55 EET 2006
	Added an SID (C64 music file) plugin.

Sat Nov 11 00:04:34 EET 2006
	Added an NSF (NES Sound Format) plugin.

Tue Apr 18 14:44:37 PDT 2006
	Added dictionaries for Finnish, French, Gaelic and Swedish
	(for printable extractors).

Thu Mar  9 17:55:09 PST 2006
	Word history extraction works (wordleaker).

Sat Feb 18 17:39:10 PST 2006
	Releasing libextractor 0.5.10.

Fri Dec 23 11:28:23 PST 2005
	Releasing libextractor 0.5.9.

Thu Jul 14 22:30:45 CEST 2005
	exiv2 works.

Sat May 21 13:42:46 EST 2005
	Releasing libextractor 0.5.0.

Fri May  6 06:02:02 EST 2005
        Added Python binding.

Thu Mar 17 08:07:04 EST 2005
	libextractor becomes a GNU package.

Thu Feb 24 00:34:15 EST 2005
	Thumbnails work.

Sun Feb 20 16:36:17 EST 2005
	Helix works (new Real format).

Sat Nov 13 13:23:23 EST 2004
	Releasing libextractor 0.3.11.

Sat Oct 23 13:21:23 EST 2004
	Releasing libextractor 0.3.10.

Fri Oct 22 19:18:38 EST 2004
	id3v2.3 and id3v2.4 work.

Sun Oct 17 18:12:11 EST 2004
	tar and tar.gz work.

Sun Oct 17 17:42:16 EST 2004
	deb works.

Sun Oct 17 13:52:25 EST 2004
	man works.

Sat Oct  2 18:02:56 EST 2004
	dvi works.

Thu Sep 23 11:25:42 EST 2004
	Added support for StarOffice and ID3v2.
	
Fri Sep 10 22:00:09 EST 2004
	Added support for RipeMD-160.

Fri Sep 10 19:49:39 EST 2004
	Added support for SHA-1 and MD5.

Fri Sep 10 10:35:27 EST 2004
	OpenOffice works.

Mon Aug 30 23:16:17 IST 2004
	ole2 works (WinWord, Excel, PowerPoint).

Wed Apr 28 19:28:39 EST 2004
	Releasing libextractor 0.3.1.

Wed Aug 25 18:59:13 IST 2004
	GNU gettext support added.

Wed Apr 28 01:26:53 EST 2004
	elf works.

Sat Apr 10 01:34:04 EST 2004
	riff / avi works.

Sat Apr 10 00:30:19 EST 2004
	mpeg works.

Sun Apr  4 20:24:39 EST 2004
	tiff works.

Wed Jul 16 13:41:34 EST 2003
	Releasing 0.2.5.

Mon Jun 30 21:27:42 EST 2003
	Releasing 0.2.4.

Thu Jan 23 00:34:20 EST 2003
	rpm works.

Tue Jan  7 18:31:41 CET 2003
	ps works.

Tue Dec 31 15:26:00 EST 2002
	pdf works.

Tue Dec 17 20:36:13 CET 2002
	MIME-type detection works.

Fri Jun  7 01:48:34 EST 2002
	real meta formats (rm, ra) work.

Tue Jun  4 23:21:38 EST 2002
	png works.

Sun Jun  2 22:49:17 EST 2002
	jpeg and HTML work. 

Sat May 18 02:35:39 EST 2002
	ogg works.

Thu May 16 00:04:03 EST 2002
	mp3 works.

Wed May 15 21:38:31 EST 2002
	Project started.


src/main/extract.c
/*
     This file is part of libextractor.
     (C) 2002, 2003, 2004, 2005, 2006, 2009, 2012 Vidyut Samanta and Christian Grothoff

     libextractor is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 3, or (at your
     option) any later version.

     libextractor is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with libextractor; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/
/**
 * @file main/extract.c
 * @brief command-line tool to run GNU libextractor
 * @author Christian Grothoff
 */
#include "platform.h"
#include "extractor.h"
#include "getopt.h"
#include <signal.h>

#define YES 1
#define NO 0


/**
 * Which keyword types should we print?
 */
static int *print;

/**
 * How verbose are we supposed to be?
 */
static int verbose;

/**
 * Run plugins in-process.
 */
static int in_process;

/**
 * Read file contents into memory, then feed them to extractor.
 */
static int from_memory;

#ifndef WINDOWS
/**
 * Install a signal handler to ignore SIGPIPE.
 */
static void
ignore_sigpipe ()
{
  struct sigaction oldsig;
  struct sigaction sig;

  memset (&sig, 0, sizeof (struct sigaction));
  sig.sa_handler = SIG_IGN;
  sigemptyset (&sig.sa_mask);
#ifdef SA_INTERRUPT
  sig.sa_flags = SA_INTERRUPT;  /* SunOS */
#else
  sig.sa_flags = SA_RESTART;
#endif
  if (0 != sigaction (SIGPIPE, &sig, &oldsig))
    fprintf (stderr,
             "Failed to install SIGPIPE handler: %s\n", strerror (errno));
}
#endif


/**
 * Information about command-line options.
 */
struct Help 
{
  /**
   * Single-character option name, '\0' for none.
   */ 
  char shortArg;
  
  /**
   * Long name of the option.
   */ 
  const char * longArg;

  /**
   * Name of the mandatory argument, NULL for no argument.
   */
  const char * mandatoryArg;

  /**
   * Help text for the option.
   */
  const char * description;
};


/**
 * Indentation for descriptions.
 */
#define BORDER 29


/**
 * Display help text (--help).
 *
 * @param general binary name
 * @param description program description
 * @param opt program options (NULL-terminated array)
 */
static void 
format_help (const char *general,
	     const char *description,
	     const struct Help *opt) 
{
  size_t slen;
  unsigned int i;
  ssize_t j;
  size_t ml;
  size_t p;
  char scp[80];
  const char *trans;
	
  printf (_("Usage: %s\n%s\n\n"),
	  gettext(general),
	  gettext(description));
  printf (_("Arguments mandatory for long options are also mandatory for short options.\n"));
  slen = 0;
  i = 0;
  while (NULL != opt[i].description) 
    {
      if (0 == opt[i].shortArg)
	printf ("      ");
      else
	printf ("  -%c, ",
		opt[i].shortArg);
      printf ("--%s",
	      opt[i].longArg);
      slen = 8 + strlen(opt[i].longArg);
      if (NULL != opt[i].mandatoryArg) 
	{
	  printf ("=%s",
		  opt[i].mandatoryArg);
	  slen += 1+strlen(opt[i].mandatoryArg);
	}
      if (slen > BORDER) 
	{
	  printf ("\n%*s", BORDER, "");
	  slen = BORDER;
	}
      if (slen < BORDER) 
	{
	  printf ("%*s", (int) (BORDER - slen), "");
	  slen = BORDER;
	}
      trans = gettext(opt[i].description);
      ml = strlen(trans);
      p = 0;
    OUTER:
      while (ml - p > 78 - slen) 
	{
	  for (j=p+78-slen;j>p;j--)
	    {
	      if (isspace( (unsigned char) trans[j]))
		{
		  memcpy(scp,
			 &trans[p],
			 j-p);
		  scp[j-p] = '\0';
		  printf ("%s\n%*s",
			  scp,
			  BORDER + 2,
			  "");
		  p = j+1;
		  slen = BORDER + 2;
		  goto OUTER;
		}
	    }
	  /* could not find space to break line */
	  memcpy (scp,
		  &trans[p],
		  78 - slen);
	  scp[78 - slen] = '\0';
	  printf ("%s\n%*s",
		  scp,
		  BORDER+2,
		  "");	
	  slen = BORDER+2;
	  p = p + 78 - slen;
	}
      /* print rest */
      if (p < ml)
	printf("%s\n",
	       &trans[p]);
      i++;
    }
}


/**
 * Run --help.
 */
static void
print_help ()
{
  static struct Help help[] = 
    {
      { 'b', "bibtex", NULL,
	gettext_noop("print output in bibtex format") },
      { 'g', "grep-friendly", NULL,
	gettext_noop("produce grep-friendly output (all results on one line per file)") },
      { 'h', "help", NULL,
	gettext_noop("print this help") },
      { 'i', "in-process", NULL,
	gettext_noop("run plugins in-process (simplifies debugging)") },
      { 'm', "from-memory", NULL,
	gettext_noop("read data from file into memory and extract from memory") },
      { 'l', "library", "LIBRARY",
	gettext_noop("load an extractor plugin named LIBRARY") },
      { 'L', "list", NULL,
	gettext_noop("list all keyword types") },
      { 'n', "nodefault", NULL,
	gettext_noop("do not use the default set of extractor plugins") },
      { 'p', "print", "TYPE",
	gettext_noop("print only keywords of the given TYPE (use -L to get a list)") },
      { 'v', "version", NULL,
	gettext_noop("print the version number") },
      { 'V', "verbose", NULL,
	gettext_noop("be verbose") },
      { 'x', "exclude", "TYPE",
	gettext_noop("do not print keywords of the given TYPE") },
      { 0, NULL, NULL, NULL },
    };
  format_help (_("extract [OPTIONS] [FILENAME]*"),
	      _("Extract metadata from files."),
	      help);

}

#include "iconv.c"

/**
 * Print a keyword list to a file.
 *
 * @param cls closure, not used
 * @param plugin_name name of the plugin that produced this value;
 *        special values can be used (i.e. '<zlib>' for zlib being
 *        used in the main libextractor library and yielding
 *        meta data).
 * @param type libextractor-type describing the meta data
 * @param format basic format information about data 
 * @param data_mime_type mime-type of data (not of the original file);
 *        can be NULL (if mime-type is not known)
 * @param data actual meta-data found
 * @param data_len number of bytes in data
 * @return 0 to continue extracting, 1 to abort
 */ 
static int
print_selected_keywords (void *cls,
			 const char *plugin_name,
			 enum EXTRACTOR_MetaType type,
			 enum EXTRACTOR_MetaFormat format,
			 const char *data_mime_type,
			 const char *data,
			 size_t data_len)
{ 
  char *keyword;
  iconv_t cd;
  const char *stype;
  const char *mt;

  if (YES != print[type])
    return 0;
  if (verbose > 3)
    fprintf (stdout,
	     _("Found by `%s' plugin:\n"),
	     plugin_name);
  mt = EXTRACTOR_metatype_to_string (type);
  stype = (NULL == mt) ? _("unknown") : gettext(mt);
  switch (format)
    {
    case EXTRACTOR_METAFORMAT_UNKNOWN:
      fprintf (stdout,
	       _("%s - (unknown, %u bytes)\n"),
	       stype,
	       (unsigned int) data_len);
      break;
    case EXTRACTOR_METAFORMAT_UTF8:
      cd = iconv_open (nl_langinfo(CODESET), "UTF-8");
      if (((iconv_t) -1) != cd)
	keyword = iconv_helper (cd,
				data,
				data_len);
      else
	keyword = strdup (data);
      if (NULL != keyword)
	{
	  fprintf (stdout,
		   "%s - %s\n",
		   stype,
		   keyword);
	  free (keyword);
	}
      if (((iconv_t) -1) != cd)
	iconv_close (cd);
      break;
    case EXTRACTOR_METAFORMAT_BINARY:
      fprintf (stdout,
	       _("%s - (binary, %u bytes)\n"),
	       stype,
	       (unsigned int) data_len);
      break;
    case EXTRACTOR_METAFORMAT_C_STRING:
      fprintf (stdout,
	       "%s - %.*s\n",
	       stype,
	       (int) data_len,
	       data);
      break;
    default:
      break;
    }
  return 0;
}


/**
 * Print a keyword list to a file without new lines.
 *
 * @param cls closure, not used
 * @param plugin_name name of the plugin that produced this value;
 *        special values can be used (i.e. '<zlib>' for zlib being
 *        used in the main libextractor library and yielding
 *        meta data).
 * @param type libextractor-type describing the meta data
 * @param format basic format information about data 
 * @param data_mime_type mime-type of data (not of the original file);
 *        can be NULL (if mime-type is not known)
 * @param data actual meta-data found
 * @param data_len number of bytes in data
 * @return 0 to continue extracting, 1 to abort
 */ 
static int
print_selected_keywords_grep_friendly (void *cls,
				       const char *plugin_name,
				       enum EXTRACTOR_MetaType type,
				       enum EXTRACTOR_MetaFormat format,
				       const char *data_mime_type,
				       const char *data,
				       size_t data_len)
{ 
  char *keyword;
  iconv_t cd;
  const char *mt;

  if (YES != print[type])
    return 0;
  mt = EXTRACTOR_metatype_to_string (type);
  if (NULL == mt)
    mt = gettext_noop ("unknown");
  switch (format)
    {
    case EXTRACTOR_METAFORMAT_UNKNOWN:      
      break;
    case EXTRACTOR_METAFORMAT_UTF8:
      if (verbose > 1)
	fprintf (stdout,
		 "%s: ",
		 gettext(mt));
      cd = iconv_open (nl_langinfo (CODESET), "UTF-8");
      if (((iconv_t) -1) != cd)
	keyword = iconv_helper (cd,
				data,
				data_len);
      else
	keyword = strdup (data);
      if (NULL != keyword)
	{
	  fprintf (stdout,
		   "`%s' ",
		   keyword);
	  free (keyword);
	}
      if (((iconv_t) -1) != cd)
	iconv_close (cd);
      break;
    case EXTRACTOR_METAFORMAT_BINARY:
      break;
    case EXTRACTOR_METAFORMAT_C_STRING:
      if (verbose > 1)
	fprintf (stdout,
		 "%s ",
		 gettext(mt));
      fprintf (stdout,
	       "`%s'",
	       data);
      break;
    default:
      break;
    }
  return 0;
}


/**
 * Entry in the map we construct for each file.
 */
struct BibTexMap
{
  /**
   * Name in bibTeX
   */
  const char *bibTexName;

  /**
   * Meta type for the value.
   */
  enum EXTRACTOR_MetaType le_type;

  /**
   * The value itself.
   */
  char *value;
};


/**
 * Type of the entry for bibtex.
 */
static char *entry_type;

/**
 * Mapping between bibTeX strings, libextractor
 * meta data types and values for the current document.
 */
static struct BibTexMap btm[] =
  {
    { "title", EXTRACTOR_METATYPE_TITLE, NULL},
    { "year", EXTRACTOR_METATYPE_PUBLICATION_YEAR, NULL },
    { "author", EXTRACTOR_METATYPE_AUTHOR_NAME, NULL },
    { "book", EXTRACTOR_METATYPE_BOOK_TITLE, NULL},
    { "edition", EXTRACTOR_METATYPE_BOOK_EDITION, NULL},
    { "chapter", EXTRACTOR_METATYPE_BOOK_CHAPTER_NUMBER, NULL},
    { "journal", EXTRACTOR_METATYPE_JOURNAL_NAME, NULL},
    { "volume", EXTRACTOR_METATYPE_JOURNAL_VOLUME, NULL},
    { "number", EXTRACTOR_METATYPE_JOURNAL_NUMBER, NULL},
    { "pages", EXTRACTOR_METATYPE_PAGE_COUNT, NULL },
    { "pages", EXTRACTOR_METATYPE_PAGE_RANGE, NULL },
    { "school", EXTRACTOR_METATYPE_AUTHOR_INSTITUTION, NULL},
    { "publisher", EXTRACTOR_METATYPE_PUBLISHER, NULL },
    { "address", EXTRACTOR_METATYPE_PUBLISHER_ADDRESS, NULL },
    { "institution", EXTRACTOR_METATYPE_PUBLISHER_INSTITUTION, NULL },
    { "series", EXTRACTOR_METATYPE_PUBLISHER_SERIES, NULL},
    { "month", EXTRACTOR_METATYPE_PUBLICATION_MONTH, NULL },
    { "url", EXTRACTOR_METATYPE_URL, NULL}, 
    { "note", EXTRACTOR_METATYPE_COMMENT, NULL},
    { "eprint", EXTRACTOR_METATYPE_BIBTEX_EPRINT, NULL },
    { "type", EXTRACTOR_METATYPE_PUBLICATION_TYPE, NULL },
    { NULL, 0, NULL }
  };


/**
 * Clean up the bibtex processor in preparation for the next round.
 */
static void 
cleanup_bibtex ()
{
  unsigned int i;
  
  for (i = 0; NULL != btm[i].bibTexName; i++)
    {
      free (btm[i].value);
      btm[i].value = NULL;
    }
  free (entry_type);
  entry_type = NULL;
}


/**
 * Callback function for printing meta data in bibtex format.
 *
 * @param cls closure, not used
 * @param plugin_name name of the plugin that produced this value;
 *        special values can be used (i.e. '<zlib>' for zlib being
 *        used in the main libextractor library and yielding
 *        meta data).
 * @param type libextractor-type describing the meta data
 * @param format basic format information about data 
 * @param data_mime_type mime-type of data (not of the original file);
 *        can be NULL (if mime-type is not known)
 * @param data actual meta-data found
 * @param data_len number of bytes in data
 * @return 0 to continue extracting (always)
 */
static int
print_bibtex (void *cls,
	      const char *plugin_name,
	      enum EXTRACTOR_MetaType type,
	      enum EXTRACTOR_MetaFormat format,
	      const char *data_mime_type,
	      const char *data,
	      size_t data_len)
{
  unsigned int i;

  if (YES != print[type])
    return 0;
  if (EXTRACTOR_METAFORMAT_UTF8 != format)
    return 0;
  if (EXTRACTOR_METATYPE_BIBTEX_ENTRY_TYPE == type)
    {
      entry_type = strdup (data);
      return 0;
    }
  for (i = 0; NULL != btm[i].bibTexName; i++)
    if ( (NULL == btm[i].value) &&
	 (btm[i].le_type == type) )
      btm[i].value = strdup (data);
  return 0;
}


/**
 * Print the computed bibTeX entry.
 *
 * @param fn file for which the entry was created.
 */
static void
finish_bibtex (const char *fn)
{
  unsigned int i;
  ssize_t n;
  const char *et;
  char temp[20];

  if (NULL != entry_type)
    et = entry_type;
  else
    et = "misc";
  if ( (NULL == btm[0].value) ||
       (NULL == btm[1].value) ||
       (NULL == btm[2].value) )          
    fprintf (stdout,
	     "@%s %s { ",
	     et,
	     fn);
  else
    {
      snprintf (temp,
		sizeof (temp),
		"%.5s%.5s%.5s",
		btm[2].value,
		btm[1].value,
		btm[0].value);
      for (n=strlen (temp)-1;n>=0;n-- )
	if (! isalnum ( (unsigned char) temp[n]) ) 
	  temp[n] = '_';
	else 
	  temp[n] = tolower ( (unsigned char) temp[n]);
      fprintf (stdout,
	       "@%s %s { ",
	       et,
	       temp);
    }
  for (i=0; NULL != btm[i].bibTexName; i++)
    if (NULL != btm[i].value) 
      fprintf (stdout,
	       "\t%s = {%s},\n",
	       btm[i].bibTexName,
	       btm[i].value);
  fprintf (stdout, "}\n\n");
}


/**
 * Main function for the 'extract' tool.  Invoke with a list of
 * filenames to extract keywords from.
 *
 * @param argc number of arguments in argv
 * @param argv command line options and filename to run on
 * @return 0 on success
 */
int
main (int argc, char *argv[])
{
  unsigned int i;
  struct EXTRACTOR_PluginList *plugins;
  int option_index;
  int c;
  char *libraries = NULL;
  int nodefault = NO;
  int defaultAll = YES;
  int bibtex = NO;
  int grepfriendly = NO;
  int ret = 0;
  EXTRACTOR_MetaDataProcessor processor = NULL;

#if ENABLE_NLS
  setlocale(LC_ALL, "");
  textdomain(PACKAGE);
#endif
#ifndef WINDOWS
  ignore_sigpipe ();
#endif
  if (NULL == (print = malloc (sizeof (int) * EXTRACTOR_metatype_get_max ())))
    {
      fprintf (stderr, 
	       "malloc failed: %s\n",
	       strerror (errno));
      return 1;
    }
  for (i = 0; i < EXTRACTOR_metatype_get_max (); i++)
    print[i] = YES;		/* default: print everything */

  while (1)
    {
      static struct option long_options[] = {
	{"bibtex", 0, 0, 'b'},
	{"grep-friendly", 0, 0, 'g'},
	{"help", 0, 0, 'h'},
	{"in-process", 0, 0, 'i'},
        {"from-memory", 0, 0, 'm'},
	{"list", 0, 0, 'L'},
	{"library", 1, 0, 'l'},
	{"nodefault", 0, 0, 'n'},
	{"print", 1, 0, 'p'},
	{"verbose", 0, 0, 'V'},
	{"version", 0, 0, 'v'},
	{"exclude", 1, 0, 'x'},
	{0, 0, 0, 0}
      };
      option_index = 0;
      c = getopt_long (argc,
		       argv, 
		       "abghiml:Lnp:vVx:",
		       long_options,
		       &option_index);

      if (c == -1)
	break;			/* No more flags to process */
      switch (c)
	{
	case 'b':
	  bibtex = YES;
	  if (NULL != processor)
	    {
	      fprintf (stderr,
		       _("Illegal combination of options, cannot combine multiple styles of printing.\n"));
	      return 0;
	    }
	  processor = &print_bibtex;
	  break;
	case 'g':
	  grepfriendly = YES;
	  if (NULL != processor)
	    {
	      fprintf (stderr,
		       _("Illegal combination of options, cannot combine multiple styles of printing.\n"));
	      return 0;
	    }
	  processor = &print_selected_keywords_grep_friendly;
	  break;
	case 'h':
	  print_help ();
	  return 0;
	case 'i':
	  in_process = YES;
	  break;
        case 'm':
          from_memory = YES;
          break;
	case 'l':
	  libraries = optarg;
	  break;
	case 'L':
	  i = 0;
	  while (NULL != EXTRACTOR_metatype_to_string (i))
	    printf ("%s\n",
		    gettext(EXTRACTOR_metatype_to_string (i++)));
	  return 0;
	case 'n':
	  nodefault = YES;
	  break;
	case 'p':
	  if (NULL == optarg) 
	    {
	      fprintf(stderr,
		      _("You must specify an argument for the `%s' option (option ignored).\n"),
		      "-p");
	      break;
	    }
	  if (YES == defaultAll)
	    {
	      defaultAll = NO;
	      i = 0;
	      while (NULL != EXTRACTOR_metatype_to_string (i))
		print[i++] = NO;
	    }
	  i = 0;
	  while (NULL != EXTRACTOR_metatype_to_string (i))
	    {
	      if ( (0 == strcmp (optarg, 
				 EXTRACTOR_metatype_to_string (i))) ||
		   (0 == strcmp (optarg, 
				 gettext(EXTRACTOR_metatype_to_string (i)))) )
		
		{
		  print[i] = YES;
		  break;
		}
	      i++;
	    }
	  if (NULL == EXTRACTOR_metatype_to_string (i))
	    {
	      fprintf(stderr,
		      "Unknown keyword type `%s', use option `%s' to get a list.\n",
		      optarg,
		       "-L");
	      return -1;
	    }
	  break;
       	case 'v':
	  printf ("extract v%s\n", PACKAGE_VERSION);
	  return 0;
	case 'V':
	  verbose++;
	  break;
	case 'x':
	  i = 0;
	  while (NULL != EXTRACTOR_metatype_to_string (i))
	    {
	      if ( (0 == strcmp (optarg, 
				 EXTRACTOR_metatype_to_string (i))) ||
		   (0 == strcmp (optarg, 
				 gettext(EXTRACTOR_metatype_to_string (i)))) )
		{
		  print[i] = NO;
		  break;
		}
	      i++;
	    }
	  if (NULL == EXTRACTOR_metatype_to_string (i))
	    {
	      fprintf (stderr,
		       "Unknown keyword type `%s', use option `%s' to get a list.\n",
		       optarg,
		       "-L");
	      return -1;
	    }
	  break;
	default:
	  fprintf (stderr,
		   _("Use --help to get a list of options.\n"));
	  return -1;
	}			/* end of parsing commandline */
    }				/* while (1) */
  if (optind < 0)
    {
      fprintf (stderr,
	       "Unknown error parsing options\n");
      free (print);
      return -1;
    }
  if (argc - optind < 1)
    {
      fprintf (stderr,
	       "Invoke with list of filenames to extract keywords form!\n");
      free (print);
      return -1;
    }

  /* build list of libraries */
  if (NO == nodefault)
    plugins = EXTRACTOR_plugin_add_defaults (in_process
					     ? EXTRACTOR_OPTION_IN_PROCESS
					     : EXTRACTOR_OPTION_DEFAULT_POLICY);
  else
    plugins = NULL;
  if (NULL != libraries)
    plugins = EXTRACTOR_plugin_add_config (plugins, 
					   libraries,
					   in_process
					   ? EXTRACTOR_OPTION_IN_PROCESS
					   : EXTRACTOR_OPTION_DEFAULT_POLICY);
  if (NULL == processor)
    processor = &print_selected_keywords;

  /* extract keywords */
  if (YES == bibtex)
    fprintf(stdout,
	    _("%% BiBTeX file\n"));
  for (i = optind; i < argc; i++) 
    {
      errno = 0;
      if (YES == grepfriendly)
	fprintf (stdout, "%s ", argv[i]);
      else if (NO == bibtex)
	fprintf (stdout,
		 _("Keywords for file %s:\n"),
		 argv[i]);
      else
	cleanup_bibtex ();
      if (NO == from_memory)
	EXTRACTOR_extract (plugins,
			   argv[i],
			   NULL, 0,
			   processor,
			   NULL);
      else
	{
	  struct stat sb;
	  unsigned char *data = NULL;
	  int f = open (argv[i], O_RDONLY
#if WINDOWS
			| O_BINARY
#endif
			);
	  if ( (-1 != f) &&
	       (0 == fstat (f, &sb)) &&
	       (NULL != (data = malloc ((size_t) sb.st_size))) &&
	       (sb.st_size == read (f, data, (size_t) sb.st_size) ) )
	    {
	      EXTRACTOR_extract (plugins,
				 NULL,
				 data, sb.st_size,
				 processor,
				 NULL);
	    }
	  else
	    {
	      if (verbose > 0) 
		fprintf(stderr,
			"%s: %s: %s\n",
			argv[0], argv[i], strerror(errno));
	      ret = 1;
	    }
	  if (NULL != data)
	    free (data);
	  if (-1 != f)
	    (void) close (f);
	}
      if (YES == grepfriendly)
	fprintf (stdout, "\n");
      continue;
    }
  if (YES == grepfriendly)
    fprintf (stdout, "\n");
  if (bibtex)
    finish_bibtex (argv[i]);
  if (verbose > 0)
    printf ("\n");
  free (print);
  EXTRACTOR_plugin_remove_all (plugins);
  plugins = NULL;
  cleanup_bibtex (); /* actually free's stuff */
  return ret;
}

/* end of extract.c */
